lapce-data/src/config.rs

```diff
@@ -988,23 +1002,13 @@ impl notify::EventHandler for ConfigWatcher {
         if let Ok(event) = event {
             match event.kind {
                 notify::EventKind::Create(_)
-                | notify::EventKind::Modify(ModifyKind::Data(DataChange::Content))
+                | notify::EventKind::Modify(_)
                 | notify::EventKind::Remove(_) => {
-                    *self.delay_handler.lock() = Some(());
```

Comment:
Do you think this might be a simpler solution? 
```
                    static WAITING: AtomicBool = AtomicBool::new(false);
                    if WAITING.load(std::sync::atomic::Ordering::Acquire) {
                        // still waiting for the config reload timeout, don't do anything
                        return;
                    }
                    WAITING.store(true, std::sync::atomic::Ordering::Release);
                    let event_sink = self.event_sink.clone();
                    std::thread::spawn(move || {
                        std::thread::sleep(std::time::Duration::from_millis(500));
                        WAITING.store(false, std::sync::atomic::Ordering::Release);
                        let _ = event_sink.submit_command(
                            LAPCE_UI_COMMAND,
                            LapceUICommand::ReloadConfig,
                            Target::Auto,
                        );
                    });
```

Reply:
3 problems:

1. There are multiple atomic read/writes here. This should be done through something like `compare_swap_exchange`.
1. This,
```
let _ = event_sink.submit_command(
                            LAPCE_UI_COMMAND,
                            LapceUICommand::ReloadConfig,
                            Target::Auto,
                        );
```
is simply putting the command to a queue. Putting the lock here is quite meaningless because simply pushing into the queue is quite fast, so this will flood the commands queue. It is possible that 1 render pass (or whatever) will contain multiple commands to reload config. So we need a way for the reload config command itself to stop any incoming command until it is done.
2. If we do prevent the flood of commands mentioned above, we still have the problem of changes that happening _while_ we are reloading config won't be notified.

```
|_____|_____________________________|____________________|___________________________
^     ^                             ^                    ^
|     |                             |                    |
|     reload config job started     |                    reload config job done but partial
first event here                    another event here (will get ignored because it failed the lock)
```

this reload config will be partial because there are changes made while the reload config job started that won't be notified.
So we need a way for the event to queue _another_ command if the command is being executed. I realized that this is basically having a channel of capacity 2.

Reply:
1. Maybe we could use a Mutex here which can also help 3
2. One thread will only spawn every 500 ms
3. We set waiting to false before submitting ReloadConfig, so it's guaranteed that a new file event will trigger a new ReloadConfig. Again, a Mutex here could be better.

Reply:
Ahh I see what you mean. I will try doing that.

Reply:
Something like this?


```
diff --git a/lapce-data/src/config.rs b/lapce-data/src/config.rs
index ee402501..4fa87fd4 100644
--- a/lapce-data/src/config.rs
+++ b/lapce-data/src/config.rs
@@ -1,7 +1,7 @@
 use std::{
     collections::HashMap,
     path::{Path, PathBuf},
-    sync::Arc,
+    sync::{Arc, Mutex},
 };
 
 use druid::{
@@ -13,9 +13,8 @@
 use lapce_proxy::plugin::wasi::find_all_volts;
 use lapce_rpc::plugin::VoltID;
 use lsp_types::{CompletionItemKind, SymbolKind};
-use notify::event::{DataChange, ModifyKind};
 use once_cell::sync::Lazy;
-use parking_lot::{Mutex, RwLock};
+use parking_lot::RwLock;
 use serde::{Deserialize, Serialize};
 use structdesc::FieldNames;
 use thiserror::Error;
@@ -971,15 +970,19 @@ pub fn get_dropdown_info(&self, kind: &str, key: &str) -> Option<DropdownInfo> {
 
 pub struct ConfigWatcher {
     event_sink: ExtEventSink,
-    delay_handler: Arc<Mutex<Option<()>>>,
+    config_mutex: Arc<Mutex<u8>>,
 }
 
 impl ConfigWatcher {
-    pub fn new(event_sink: ExtEventSink) -> Self {
-        Self {
-            event_sink,
-            delay_handler: Arc::new(Mutex::new(None)),
-        }
+    pub fn new(event_sink: ExtEventSink) -> (Self, Arc<Mutex<u8>>) {
+        let config_mutex = Arc::new(Mutex::new(0));
+        (
+            Self {
+                event_sink,
+                config_mutex: config_mutex.clone(),
+            },
+            config_mutex,
+        )
     }
 }
 
@@ -988,23 +991,19 @@ fn handle_event(&mut self, event: notify::Result<notify::Event>) {
         if let Ok(event) = event {
             match event.kind {
                 notify::EventKind::Create(_)
-                | notify::EventKind::Modify(ModifyKind::Data(DataChange::Content))
+                | notify::EventKind::Modify(_)
                 | notify::EventKind::Remove(_) => {
-                    *self.delay_handler.lock() = Some(());
-                    let delay_handler = self.delay_handler.clone();
-                    let event_sink = self.event_sink.clone();
-                    std::thread::spawn(move || {
-                        std::thread::sleep(std::time::Duration::from_millis(500));
-                        if delay_handler.lock().take().is_some() {
-                            let _ = event_sink.submit_command(
-                                LAPCE_UI_COMMAND,
-                                LapceUICommand::ReloadConfig,
-                                Target::Auto,
-                            );
-                        }
-                    });
+                    let mut count = self.config_mutex.lock().unwrap();
+                    if *count < 2 {
+                        let _ = self.event_sink.submit_command(
+                            LAPCE_UI_COMMAND,
+                            LapceUICommand::ReloadConfig,
+                            Target::Auto,
+                        );
+                        *count = *count + 1;
+                    }
                 }
-                _ => (),
+                _ => {}
             }
         }
     }
diff --git a/lapce-data/src/data.rs b/lapce-data/src/data.rs
index eacea81b..7ea1f962 100644
--- a/lapce-data/src/data.rs
+++ b/lapce-data/src/data.rs
@@ -7,7 +7,7 @@
     io::{BufReader, Read, Write},
     path::{Path, PathBuf},
     rc::Rc,
-    sync::Arc,
+    sync::{Arc, Mutex},
     thread,
     time::Instant,
 };
@@ -435,6 +435,7 @@ pub struct LapceWindowData {
     pub update_in_progress: bool,
     #[data(ignore)]
     pub log_file: Arc<Option<PathBuf>>,
+    pub config_mutex: Arc<Mutex<u8>>,
 }
 
 impl LapceWindowData {
@@ -513,8 +514,8 @@ pub fn new(
             Target::Widget(active_tab_id),
         );
 
-        let mut watcher =
-            notify::recommended_watcher(ConfigWatcher::new(event_sink)).unwrap();
+        let (config_watcher, config_mutex) = ConfigWatcher::new(event_sink);
+        let mut watcher = notify::recommended_watcher(config_watcher).unwrap();
         if let Some(path) = LapceConfig::settings_file() {
             let _ = watcher.watch(&path, notify::RecursiveMode::Recursive);
         }
@@ -545,6 +546,7 @@ pub fn new(
             latest_release,
             update_in_progress,
             log_file,
+            config_mutex,
         }
     }
 
diff --git a/lapce-ui/src/window.rs b/lapce-ui/src/window.rs
index 296bc332..6c26728a 100644
--- a/lapce-ui/src/window.rs
+++ b/lapce-ui/src/window.rs
@@ -352,6 +352,15 @@ fn event(
                         }
                         Arc::make_mut(&mut data.keypress)
                             .update_keymaps(&data.config);
+                        let mut count = data.config_mutex.lock().unwrap();
+                        *count = *count - 1;
+                        if *count > 0 {
+                            ctx.submit_command(Command::new(
+                                LAPCE_UI_COMMAND,
+                                LapceUICommand::ReloadConfig,
+                                Target::Auto,
+                            ));
+                        }
                         ctx.set_handled();
                     }
                    LapceUICommand::ReloadWindow => {
```

Reply:
We don't need to check it in the main thread.

Reply:
Ahh but then we will spawn thousands of thread all just to lock hang on a mutex. This is what causing the hang when you download/delete plugins AFAIK.

Reply:
Is this okay? It's basically what my original PR did but with some additional heurestics:

1. If changes are in settings.toml or keymaps.toml, submit command immediately.
2. Otherwise, lock bool, sleep for 2 seconds, submit command and unlock bool again.

```
diff --git a/lapce-data/src/config.rs b/lapce-data/src/config.rs
index ee402501..fd109a26 100644
--- a/lapce-data/src/config.rs
+++ b/lapce-data/src/config.rs
@@ -1,7 +1,7 @@
 use std::{
     collections::HashMap,
     path::{Path, PathBuf},
-    sync::Arc,
+    sync::{atomic::AtomicBool, Arc},
 };
 
 use druid::{
@@ -13,9 +13,8 @@
 use lapce_proxy::plugin::wasi::find_all_volts;
 use lapce_rpc::plugin::VoltID;
 use lsp_types::{CompletionItemKind, SymbolKind};
-use notify::event::{DataChange, ModifyKind};
 use once_cell::sync::Lazy;
-use parking_lot::{Mutex, RwLock};
+use parking_lot::RwLock;
 use serde::{Deserialize, Serialize};
 use structdesc::FieldNames;
 use thiserror::Error;
@@ -971,14 +970,14 @@ pub fn get_dropdown_info(&self, kind: &str, key: &str) -> Option<DropdownInfo> {
 
 pub struct ConfigWatcher {
     event_sink: ExtEventSink,
-    delay_handler: Arc<Mutex<Option<()>>>,
+    delay_handler: Arc<AtomicBool>,
 }
 
 impl ConfigWatcher {
     pub fn new(event_sink: ExtEventSink) -> Self {
         Self {
             event_sink,
-            delay_handler: Arc::new(Mutex::new(None)),
+            delay_handler: Arc::new(AtomicBool::new(false)),
         }
     }
 }
@@ -988,23 +987,51 @@ fn handle_event(&mut self, event: notify::Result<notify::Event>) {
         if let Ok(event) = event {
             match event.kind {
                 notify::EventKind::Create(_)
-                | notify::EventKind::Modify(ModifyKind::Data(DataChange::Content))
+                | notify::EventKind::Modify(_)
                 | notify::EventKind::Remove(_) => {
-                    *self.delay_handler.lock() = Some(());
-                    let delay_handler = self.delay_handler.clone();
-                    let event_sink = self.event_sink.clone();
-                    std::thread::spawn(move || {
-                        std::thread::sleep(std::time::Duration::from_millis(500));
-                        if delay_handler.lock().take().is_some() {
-                            let _ = event_sink.submit_command(
-                                LAPCE_UI_COMMAND,
-                                LapceUICommand::ReloadConfig,
-                                Target::Auto,
-                            );
+                    let short = event
+                        .paths
+                        .iter()
+                        .filter_map(|p| p.file_name())
+                        .any(|p| p == "settings.toml" || p == "keymaps.toml");
+
+                    if short {
+                        let _ = self.event_sink.submit_command(
+                            LAPCE_UI_COMMAND,
+                            LapceUICommand::ReloadConfig,
+                            Target::Auto,
+                        );
+                    } else {
+                        if self
+                            .delay_handler
+                            .compare_exchange(
+                                false,
+                                true,
+                                std::sync::atomic::Ordering::Relaxed,
+                                std::sync::atomic::Ordering::Relaxed,
+                            )
+                            .is_ok()
+                        {
+                            let config_mutex = self.delay_handler.clone();
+                            let event_sink = self.event_sink.clone();
+                            std::thread::spawn(move || {
+                                std::thread::sleep(std::time::Duration::from_secs(
+                                    2,
+                                ));
+                                let _ = event_sink.submit_command(
+                                    LAPCE_UI_COMMAND,
+                                    LapceUICommand::ReloadConfig,
+                                    Target::Auto,
+                                );
+                                config_mutex.store(
+                                    false,
+                                    std::sync::atomic::Ordering::Relaxed,
+                                );
+                            });
                         }
-                    });
+                    }
                 }
-                _ => (),
+                _ => {}
             }
         }
     }
```